doc/developer-notes.md

```diff
@@ -115,6 +115,9 @@ code.
     Use `reinterpret_cast` and `const_cast` as appropriate.
   - Prefer [`list initialization ({})`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list) where possible.
     For example `int x{0};` instead of `int x = 0;` or `int x(0);`
+  - Recursive functions are generally discouraged, and checked for by
+    clang-tidy. When recursiveness is the best approach, use
```

Comment:
Not sure if the point is to discourage recursion, or to tell developers that it has to be marked explicitly? I think we can consider it implied that people should use recursion only when appropriate, just like any other practices that can lead to bugs. And disagreements about whether or not [fill in the blank] is appropriate can be resolved in review on a case-by-case basis.
```suggestion
  - Recursion is checked by clang-tidy and thus must be made explicit. Use
```

Reply:
That framing makes a lot more sense to me 👍